File: openedu/week9/lab9_2.cpp
#include "edx-io.hpp"
#include <iostream>
#include <fstream>
#include <vector>
#include <map>
#include <algorithm>
#include <string>
#include <list>
#include <assert.h>
#include <tuple>

using namespace std;
typedef unsigned long long ull;

struct CharEntry {
	ull left, right;
	CharEntry(ull left = 0, ull right = 0) : left(left), right(right) {}
};

int main() {
	ifstream fin("input.txt");

	string s;
	char c;
	while (fin >> c) {
		if (c != ' ') {
			s.push_back(c);
		}
	}
	
	// counting chars
	map<char, CharEntry*> right, left;
	for (auto cc : s) {
		auto ce = right.find(cc);
		if (ce != right.end()) {
			ce->second->right++;
		} else {
			right[cc]= new CharEntry(0, 1);
		}
	}
	
	ull output = 0;
	c = s[0];
	auto ce = right[c];
	left[c] = ce;
	ce->left++;
	ce->right--;

	for (auto cc : s.substr(1)) {
		auto ce = right[cc];
		ce->right--;

		for (auto le : left) {
			output += le.second->left * le.second->right;
		}

		ce->left++;
		left[cc] = ce;
	}

	io << output;
	return 0;
}

File: openedu/week9/lab9_2.py
from collections import Counter

with open("input.txt") as f:
    s = f.read().strip().replace(' ', '')


class Entry:
    left: int
    right: int

    def __init__(self, left=0, right=0):
        self.left = left
        self.right = right

    def __repr__(self):
        return f"L: {self.left}, R: {self.right}"


output = 0
if len(s) >= 3:
    right = {v: Entry(right=right) for v, right in Counter(s).items()}
    left = {}

    cc = s[0]
    ce = right[cc]
    left[cc] = ce
    ce.left += 1
    ce.right -= 1

    output = 0
    for cc in s[1:]:
        ce = right[cc]
        ce.right -= 1

        for k, v in left.items():
            output += v.left * v.right

        ce.left += 1
        left[cc] = ce

with open("output.txt", "w") as f:
    f.write(str(output))

